Charting improvements

hwbllmnn opened the pull request:
This adds a lot of new functionality:

* support for grouped bar charts
* lots of improvements for timeseries charts: can now do (almost?) anything that https://github.com/OpenBfS/gis-client can do
* legend rendering
* improvements of node setup (notably runs with babel v7, a dev build in `package.json` etc.)
* automated layout

It's still in an early state, but it's getting there.

@terrestris/devs Please review.

jansule commented on package.json:
```diff
@@ -1,10 +1,14 @@
   "main": "src/index.js",
   "scripts": {
-    "release": "np --no-yarn"
+    "release": "np --no-yarn",
```
You might also want to directly push tags here, so npm and github are in sync

hwbllmnn replied:
Thanks, added this in 0fc5684

jansule commented on webpack.config.js:
```diff
@@ -0,0 +1,41 @@
+  module: {
+    rules: [{
+      test: /\.jsx?$/,
+      exclude: /node_modules\/(?!@terrestris)/,
```
This should not be necessary anymore, with latest versions of `@terrestris` packages. They now also export es5 builds.

jansule replied:
Edit: I don't see any dependencies to `@terrestris` packages in your package.json. So you should check if you have to exclude the whole `/node_modules` directory or if you simply can remove the whole line.

hwbllmnn replied:
I'm leaving that in since in my experience this works just fine if we ever add any terrestris dependencies. We're also using `d3` es6 imports.

ahennr commented on src/BarComponent/BarComponent.js:
```diff
@@ -0,0 +1,203 @@
+   * @param  {d3.scale} y the y scale
+   */
+  renderUncertainty(node, x, y) {
+    node
```
may node ne `undefined`? ;)

hwbllmnn replied:
Not really, since we create all the nodes ourselves.

ahennr commented on src/LabelUtil/LabelUtil.js:
```diff
@@ -0,0 +1,85 @@
+   * @param  {Integer} lineHeight The lineHeight in em to use
+   */
+  static wordWrap(textEl, width, leftPadding, lineHeight) {
+    textEl.each(function() {
```
see above comment

hwbllmnn replied:
The worst that could happen is that there aren't any label nodes, so we'd iterate over an empty selection here.

ahennr commented on src/ScaleUtil/ScaleUtil.js:
```diff
@@ -45,13 +49,17 @@ class ScaleUtil {
       isNaN(axisDomain[0]) || isNaN(axisDomain[1]))) {
       if (axisDomain[0] === 0 || isNaN(axisDomain[0])) {
-        axisDomain[0] = Number.MIN_VALUE;
+        axisDomain[0] = 0.000001;
```
Maybe this is a candidate for constant field value defined at the beginning?

hwbllmnn replied:
Done, collecting refinements in https://github.com/terrestris/d3-util/pull/6

hwbllmnn commented on src/AxesUtil/AxesUtil.js:
```diff
@@ -50,37 +52,160 @@ class AxesUtil {
+  static createAxis(config, scale, axisFunc) {
     let tickFormatter;
-    if (type === 'time') {
+    if (config.scale === 'time') {
```
Added a check in ba2dd4c.

Unfortunately, we don't have a test suite yet, I will add one if I find the time.

hwbllmnn commented on src/BarComponent/BarComponent.js:
```diff
@@ -0,0 +1,203 @@
+   * @param  {number[]} size the size of the svg
+   */
+  render(root, size) {
+    if (!this.config.data || this.config.data.data.length === 0) {
```
Added a check in ba2dd4c

hwbllmnn commented on src/AxesUtil/AxesUtil.js:
```diff
@@ -50,37 +52,160 @@ class AxesUtil {
+   * @param  {selection} selection the d3 selection to append the axes to
+   */
+  static drawYAxis(y, config, selection) {
+    if (!config.display) {
```
Added a check in ba2dd4c

hwbllmnn commented:
Thanks for the reviews! I've added https://github.com/terrestris/d3-util/pull/6 where I've addressed the issues and will continue to refine the library.
